### PR TITLE
Update the GMRES implementation.

### PR DESCRIPTION
**Pull request by bangerth**
In particular, use the VectorMemory::Pointer class (see #4935) as well
as a few other cleanups. The commits are probably easiest to review
separately.

### CODE REVIEW

**Comment by bangerth** [comment]
Passes the testsuite on my system.

File: include/deal.II/lac/solver_gmres.h [hunk 1]
```diff
@@ -99,13 +101,7 @@ namespace internal
       /**
        * Field for storing the vectors.
        */
-      std::vector<VectorType *> data;
-
-      /**
-       * Offset of the first vector. This is for later when vector rotation
-       * will be implemented.
```
Comment (davydden) [inline include/deal.II/lac/solver_gmres.h]: i suppose it never was implemented :smile:

File: include/deal.II/lac/vector_memory.h [hunk 2]
```diff
@@ -127,13 +127,25 @@ class VectorMemory : public Subscriptor
...
-     * Constructor, automatically allocating a vector from @p mem.
+     * Default constructor. This constructor corresponds to a @p nullptr
+     * object that does not own a vector. It can, however, later be
+     * assigned another Pointer object via move assignment in which case
+     * it will steal the vector owned by the other object
+     * (as @p std::unique_ptr does).
+     */
+    Pointer() = default;
+
+    Pointer &operator = (Pointer &&) = default;
```
Comment (tjhei) [inline include/deal.II/lac/vector_memory.h]: Can you add documentation for this move assignment?

File: include/deal.II/lac/solver_gmres.h (hunk 1 above)
Comment (bangerth) [inline include/deal.II/lac/solver_gmres.h]: Nope :-)

**Comment by bangerth** [comment]
I fixed the missing documentation.

**Pull request merged**